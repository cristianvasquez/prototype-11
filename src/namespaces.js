import rdf from 'rdf-ext'

const VAULT_NAMESPACE = 'http://notes/'

const ns = {
  schema: rdf.namespace('http://schema.org/'),
  vault: rdf.namespace('http://vault.org/'),
  rdf: rdf.namespace('http://www.w3.org/1999/02/22-rdf-syntax-ns#'),
  rdfs: rdf.namespace('http://www.w3.org/2000/01/rdf-schema#'),
  this: rdf.namespace(VAULT_NAMESPACE)
}

export { ns }