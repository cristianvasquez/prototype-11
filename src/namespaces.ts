import namespace from '@rdfjs/namespace'
import {VAULT_NAMESPACE} from './config'

const ns = {
    schema: namespace('http://schema.org/'),
    vault: namespace('http://vault.org/'),
    rdf: namespace('http://www.w3.org/1999/02/22-rdf-syntax-ns#'),
    rdfs: namespace('http://www.w3.org/2000/01/rdf-schema#'),
    this: namespace(VAULT_NAMESPACE)
}

export {ns}