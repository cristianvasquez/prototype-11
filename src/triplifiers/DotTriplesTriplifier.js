import rdf from 'rdf-ext'
import { ns } from '../namespaces.js'
import 'core-js/actual/index.js'
import { getDotTriplesFromText, withEntities } from './dotTriples.js'

function rawToURI (str) {

  if ('is-a' === str.toLowerCase()) {
    return ns.rdf.type
  }

  return rdf.namedNode(`http://props/${canonical(str)}`)
}

function toTermsSP (term) {
  if (term.entities) {
    return term.entities.map(term => term.uri)
  } else {
    return [rawToURI(term.raw)]
  }
}

function toTermsO (term) {
  if (term.entities) {
    return term.entities.map(term => term.uri)
  } else {
    return [rdf.literal(term.raw)]
  }
}

function canonical (str) {
  return str.toLowerCase().replaceAll(' ', '-')
}

class DotTriplesTriplifier {

  constructor (docUri, ns) {
    if (!docUri || !docUri.termType || docUri.termType !== 'NamedNode') {
      throw Error('no document URI defined')
    }
    if (!ns) {
      throw Error('no namespaces defined')
    }
    this.docUri = docUri
    this.ns = ns
  }

  triplififyText (text, uriResolvers) {
    const quads = []
    for (const dotTriple of getDotTriplesFromText(text)) {
      const triple = withEntities(dotTriple, uriResolvers)

      if (!triple.exception) {
        for (const s of toTermsSP(triple.subject)) {
          for (const p of toTermsSP(triple.predicate)) {
            for (const o of toTermsO(triple.object)) {
              quads.push(rdf.quad(s, p, o))
            }
          }
        }
      }

    }

    return rdf.dataset().addAll(quads)
  }

}

export { DotTriplesTriplifier, canonical, rawToURI }