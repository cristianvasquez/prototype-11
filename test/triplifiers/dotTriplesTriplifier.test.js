import expect from 'expect'
import toMatchSnapshot from 'expect-mocha-snapshot'
import { DotTriplesTriplifier } from '../../src/triplifiers/DotTriplesTriplifier.js'
import { ns } from '../../src/namespaces.js'
expect.extend({ toMatchSnapshot })

const phrases = [
  'properties :: and values are separated by double colons',
  `
  to simplify things, each
  property :: value
  spans one line
  `,
  `
  a :: b
  c :: d  
  `,
  `
  a :: b
  a :: b  
  `,
  `a :: h
a :: i
a :: i`,
  'Maria :: climbed :: the tree',
  'subject :: \'Quotes escape :: characters\' :: object',
  'internal link :: [[link]]',
  'is a :: [[Editor]], [[Jour]]',
  'external link :: <http://example.org>',
  'external link :: http://example.org',
  'date :: 2022-01-07 15:07',
  'A :: B :: C :: D',
  'vaccination dates :: 2022-01-07, next week',
  '[[Note]] :: is a :: note',
  ' something :: points at :: [[Note]]',
  '__THIS__ :: is a :: note',
  ' something :: points at :: __THIS__',
]


function encodeURI (name) {
  return ns.this[btoa(unescape(encodeURIComponent(`http://${name}`)))]
}

describe('[triplififyText]', function () {

  const currentPath = 'CURRENT_NAME'
  const uriResolvers = {
    resolvePathByNoteName: (str) => `PATH_${str}`,
    resolveURIByNoteName: (str) => {
      return encodeURI(str)
    },
    getCurrentNote: () => {
      return {
        path: 'CURRENT_NAME',
        uri: encodeURI(currentPath),
        name: 'CURRENT_NAME'
      }
    },
  }

  const triplifier = new DotTriplesTriplifier(encodeURI(currentPath), ns)

  phrases.forEach((current) => {
    it(`"${current}"`, async function () {

      let actual = []
      for (const quad of triplifier.triplififyText(current, uriResolvers)) {
        actual.push(quad)
      }
      expect({
        size: actual.length,
        triples: actual,
      }).toMatchSnapshot(this)
    })
  })

})
