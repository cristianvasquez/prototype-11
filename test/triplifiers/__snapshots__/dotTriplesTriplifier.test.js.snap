// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[triplififyText] "
  a :: b
  a :: b  
  " 1`] = `
Object {
  "size": 1,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": Literal {
        "datatype": NamedNode {
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "value": "b",
      },
      "predicate": NamedNode {
        "value": "http://properties/a",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
  ],
}
`;

exports[`[triplififyText] "
  a :: b
  c :: d  
  " 1`] = `
Object {
  "size": 2,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": Literal {
        "datatype": NamedNode {
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "value": "b",
      },
      "predicate": NamedNode {
        "value": "http://properties/a",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
    Quad {
      "graph": DefaultGraph {},
      "object": Literal {
        "datatype": NamedNode {
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "value": "d",
      },
      "predicate": NamedNode {
        "value": "http://properties/c",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
  ],
}
`;

exports[`[triplififyText] "
  to simplify things, each
  property :: value
  spans one line
  " 1`] = `
Object {
  "size": 1,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": Literal {
        "datatype": NamedNode {
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "value": "value",
      },
      "predicate": NamedNode {
        "value": "http://properties/property",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
  ],
}
`;

exports[`[triplififyText] " something :: points at :: [[Note]]" 1`] = `
Object {
  "size": 1,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": NamedNode {
        "value": "http://notes/aHR0cDovL05vdGU=",
      },
      "predicate": NamedNode {
        "value": "http://properties/points-at",
      },
      "subject": NamedNode {
        "value": "http://properties/something",
      },
    },
  ],
}
`;

exports[`[triplififyText] " something :: points at :: __THIS__" 1`] = `
Object {
  "size": 1,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
      "predicate": NamedNode {
        "value": "http://properties/points-at",
      },
      "subject": NamedNode {
        "value": "http://properties/something",
      },
    },
  ],
}
`;

exports[`[triplififyText] "[[Note]] :: is a :: note" 1`] = `
Object {
  "size": 1,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": Literal {
        "datatype": NamedNode {
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "value": "note",
      },
      "predicate": NamedNode {
        "value": "http://properties/is-a",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL05vdGU=",
      },
    },
  ],
}
`;

exports[`[triplififyText] "__THIS__ :: is a :: note" 1`] = `
Object {
  "size": 1,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": Literal {
        "datatype": NamedNode {
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "value": "note",
      },
      "predicate": NamedNode {
        "value": "http://properties/is-a",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
  ],
}
`;

exports[`[triplififyText] "A :: B :: C :: D" 1`] = `
Object {
  "size": 0,
  "triples": Array [],
}
`;

exports[`[triplififyText] "Maria :: climbed :: the tree" 1`] = `
Object {
  "size": 1,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": Literal {
        "datatype": NamedNode {
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "value": "the tree",
      },
      "predicate": NamedNode {
        "value": "http://properties/climbed",
      },
      "subject": NamedNode {
        "value": "http://properties/maria",
      },
    },
  ],
}
`;

exports[`[triplififyText] "a :: h
a :: i
a :: i" 1`] = `
Object {
  "size": 2,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": Literal {
        "datatype": NamedNode {
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "value": "h",
      },
      "predicate": NamedNode {
        "value": "http://properties/a",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
    Quad {
      "graph": DefaultGraph {},
      "object": Literal {
        "datatype": NamedNode {
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "value": "i",
      },
      "predicate": NamedNode {
        "value": "http://properties/a",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
  ],
}
`;

exports[`[triplififyText] "date :: 2022-01-07 15:07" 1`] = `
Object {
  "size": 1,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": Literal {
        "datatype": NamedNode {
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "value": "2022-01-07 15:07",
      },
      "predicate": NamedNode {
        "value": "http://properties/date",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
  ],
}
`;

exports[`[triplififyText] "external link :: <http://example.org>" 1`] = `
Object {
  "size": 1,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": Literal {
        "datatype": NamedNode {
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "value": "<http://example.org>",
      },
      "predicate": NamedNode {
        "value": "http://properties/external-link",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
  ],
}
`;

exports[`[triplififyText] "external link :: http://example.org" 1`] = `
Object {
  "size": 1,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": Literal {
        "datatype": NamedNode {
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "value": "http://example.org",
      },
      "predicate": NamedNode {
        "value": "http://properties/external-link",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
  ],
}
`;

exports[`[triplififyText] "internal link :: [[link]]" 1`] = `
Object {
  "size": 1,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": NamedNode {
        "value": "http://notes/aHR0cDovL2xpbms=",
      },
      "predicate": NamedNode {
        "value": "http://properties/internal-link",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
  ],
}
`;

exports[`[triplififyText] "is a :: [[Editor]], [[Jour]]" 1`] = `
Object {
  "size": 2,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": NamedNode {
        "value": "http://notes/aHR0cDovL0VkaXRvcg==",
      },
      "predicate": NamedNode {
        "value": "http://properties/is-a",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
    Quad {
      "graph": DefaultGraph {},
      "object": NamedNode {
        "value": "http://notes/aHR0cDovL0pvdXI=",
      },
      "predicate": NamedNode {
        "value": "http://properties/is-a",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
  ],
}
`;

exports[`[triplififyText] "properties :: and values are separated by double colons" 1`] = `
Object {
  "size": 1,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": Literal {
        "datatype": NamedNode {
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "value": "and values are separated by double colons",
      },
      "predicate": NamedNode {
        "value": "http://properties/properties",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
  ],
}
`;

exports[`[triplififyText] "subject :: 'Quotes escape :: characters' :: object" 1`] = `
Object {
  "size": 0,
  "triples": Array [],
}
`;

exports[`[triplififyText] "vaccination dates :: 2022-01-07, next week" 1`] = `
Object {
  "size": 1,
  "triples": Array [
    Quad {
      "graph": DefaultGraph {},
      "object": Literal {
        "datatype": NamedNode {
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "value": "2022-01-07, next week",
      },
      "predicate": NamedNode {
        "value": "http://properties/vaccination-dates",
      },
      "subject": NamedNode {
        "value": "http://notes/aHR0cDovL0NVUlJFTlRfTkFNRQ==",
      },
    },
  ],
}
`;
